File: scala/src/test/scala/com/gildedrose/GildedRoseTest.scala
package com.gildedrose

import org.scalatest._

class GildedRoseTest extends FlatSpec with Matchers {

  it should "degrade twice as fast when SellIn date is passed" in {
    val items = Array[Item](new Item("anyItem", 0, 2))
    val app = new GildedRose(items)
    app.updateQuality()
    app.items(0).quality shouldBe(0)
  }

  it should "never set the quality to negative" in {
    val items = Array[Item](new Item("anyItem", 1, 0))
    val app = new GildedRose(items)
    app.updateQuality()
    app.items(0).quality shouldBe(0)
  }

  it should "increase the quality of Aged Brie over time" in {
    val items = Array[Item](new Item("Aged Brie", 1, 2))
    val app = new GildedRose(items)
    app.updateQuality()
    app.items(0).quality shouldBe(3)
  }

  it should "limit quality to 50" in {
    val items = Array[Item](new Item("Aged Brie", 1, 50))
    val app = new GildedRose(items)
    app.updateQuality()
    app.items(0).quality shouldBe(50)
  }

  it should "never decrease the quality of Sulfuras" in {
    val items = Array[Item](new Item("Sulfuras, Hand of Ragnaros", 1, 80))
    val app = new GildedRose(items)
    app.updateQuality()
    app.items(0).quality shouldBe(80)
  }

  it should "never have to sell Sulfuras" in {
    val items = Array[Item](new Item("Sulfuras, Hand of Ragnaros", 1, 80))
    val app = new GildedRose(items)
    app.updateQuality()
    app.items(0).sellIn shouldBe(1)
  }

  it should "increase the value of backstage passes by 2 or 3 or drop to 0" in {
    val items = Array[Item](
      new Item("Backstage passes to a TAFKAL80ETC concert", 11, 40),
      new Item("Backstage passes to a TAFKAL80ETC concert", 6, 40),
      new Item("Backstage passes to a TAFKAL80ETC concert", 3, 40),
      new Item("Backstage passes to a TAFKAL80ETC concert", 0, 40)
    )
    val app = new GildedRose(items)
    app.updateQuality()
    app.items(0).quality shouldBe(41)
    app.items(1).quality shouldBe(42)
    app.items(2).quality shouldBe(43)
    app.items(3).quality shouldBe(0)
  }
}

File: Kotlin/src/test/kotlin/com/gildedrose/GildedRoseTest.kt
package com.gildedrose
import org.junit.Assert.*
import org.junit.Test

class GildedRoseTest {

    @Test
    fun `Normal item should degrade in sell in days remaining after a daily update`() {
        val items = arrayOf(Item("Morning Glory Dew", 10, 10))
        val app = GildedRose(items)
        app.updateQuality()
        assertEquals("Morning Glory Dew", app.items[0].name)
        assertEquals(9, app.items[0].sellIn)
    }

    @Test
    fun `Normal item should degrade in quality after a daily update`() {
        val items = arrayOf(Item("Morning Glory Dew", 10, 10))
        val app = GildedRose(items)
        app.updateQuality()
        assertEquals("Morning Glory Dew", app.items[0].name)
        assertEquals(9, app.items[0].quality)
    }

    @Test
    fun `Normal item should degrade in quality after multiple daily updates`() {
        val items = arrayOf(Item("Morning Glory Dew", 10, 10))
        val app = GildedRose(items)
        for (days in 1..5) {
            app.updateQuality()
        }
        assertEquals("Morning Glory Dew", app.items[0].name)
        assertEquals(5, app.items[0].quality)
    }

    @Test
    fun `Normal item should degrade in sell in days after multiple daily updates`() {
        val items = arrayOf(Item("Morning Glory Dew", 10, 10))
        val app = GildedRose(items)
        for (days in 1..5) {
            app.updateQuality()
        }
        assertEquals("Morning Glory Dew", app.items[0].name)
        assertEquals(5, app.items[0].sellIn)
    }

    @Test
    fun `Normal item should not have negative quality after multiple days`() {
        val items = arrayOf(Item("Morning Glory Dew", 10, 2))
        val app = GildedRose(items)
        for (days in 1..5) {
            app.updateQuality()
        }
        assertEquals("Morning Glory Dew", app.items[0].name)
        assertEquals(0, app.items[0].quality)
    }

    @Test
    fun `Normal item quality decreases twice as fast after sell in date`() {
        val items = arrayOf(Item("Morning Glory Dew", 2, 20))
        val app = GildedRose(items)
        for (days in 1..5) {
            app.updateQuality()
        }
        assertEquals("Morning Glory Dew", app.items[0].name)
        assertEquals(12, app.items[0].quality)
    }

}


